lab_11.cpp

```diff
@@ -0,0 +1,211 @@
+  const int max_threshold{255};
+  int current_threshold{30};
+  bool use_otsu{false};
+  bool use_adaptive_model{false};
+  float adaptive_update_ratio{0.1f};
+
+  const std::string threshold_trackbar_name{"Threshold"};
+  cv::createTrackbar(threshold_trackbar_name, win_name_input, &current_threshold, max_threshold);
+
+  std::shared_ptr<MultivariateNormalModel> model;
+  cv::Mat current_samples;
+
+  while (true)
+  {
+    cap >> frame;
+    if (frame.empty())
+    {
+      throw std::runtime_error("The camera returned an empty frame. Is the camera ok?");
+    }
+
+    // Extract pixel features from the image.
+    cv::Mat feature_image = extractFeatures(frame);
+
+    if (model)
+    {
+      // Compute how well the pixel features fit with the model.
+      cv::Mat mahalanobis_img = model->computeMahalanobisDistances(feature_image);
+      cv::imshow(win_name_result, mahalanobis_img);
+
+      // Segment out the areas of the image that fits well enough.
+      cv::Mat segmented = performSegmentation(mahalanobis_img, current_threshold, use_otsu);
+      cv::setTrackbarPos(threshold_trackbar_name, win_name_input, current_threshold);
+
+      // Set segmented area to green.
+      frame.setTo(cv::Scalar{0, 255, 0}, segmented);
+    }
+
+    // Update if using adaptive model.
+    if (model && use_adaptive_model)
+    {
+      cv::Mat new_samples = extractTrainingSamples(feature_image, sampling_rectangle);
+      updateSamples(current_samples, new_samples, adaptive_update_ratio);
+      model = std::make_shared<MultivariateNormalModel>(current_samples);
+    }
+
+    // Draw current frame.
+    drawSamplingRectangle(frame, sampling_rectangle);
+    cv::imshow(win_name_input, frame);
+
+    // Get input from keyboard
+    char key = static_cast<char>(cv::waitKey(10));
+    if (key == ' ')
+    {
+      // Press space to train a model based on the samples in the rectangle.
+      current_samples = extractTrainingSamples(feature_image, sampling_rectangle);
+      model = std::make_shared<MultivariateNormalModel>(current_samples);
+    }
+    else if (key == 'o')
+    {
+      // Press 'o' to toggle use of Otsu's method.
+      use_otsu = !use_otsu;
+    }
+    else if (key == 'a')
+    {
+      // Press 'a' to toggle use of adaptive model.
+      use_adaptive_model = !use_adaptive_model;
+    }
+    else if (key >= 0)
+    {
+      // Press any other key to exit the lab.
+      break;
+    }
+  }
+}
+
+
+void updateSamples(cv::Mat& old_samples, const cv::Mat& new_samples, float update_ratio)
+{
+  // Draw uniformly distributed random numbers
+  cv::Mat rand_num = cv::Mat::zeros(1,new_samples.cols,CV_32FC1);
+  cv::randu(rand_num,0.,1.);
+
+  // Update samples
+  for (int i = 0; i<rand_num.cols; i++)
+  {
+    if (rand_num.at<float>(0,i) < update_ratio)
+    {
+      new_samples.col(i).copyTo(old_samples.col(i));
+    }
+  }
+}
+
+
+cv::Mat extractFeatures(const cv::Mat& frame)
+{
+  cv::Mat feature_image = frame.clone();
+  return feature_image;
+}
```

Comment:
🤔 Nå har ikke jeg sett på laben enda, men er dette stedet hvor de for eksempel skal endre fargerom? Ellers så var det en litt rar funksjon...

Reply:
Akkurat nå er den ikke så spennende, men det er vel her det er naturlig å utvide feature vektorene med flere features?